Sources/NIOWebSocket/NIOWebSocketClientUpgrader.swift

```diff
@@ -0,0 +1,88 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2019 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+import NIOHTTP1
+
+/// A `HTTPClientProtocolUpgrader` that knows how to do the WebSocket upgrade dance.
+///
+/// This upgrader assumes that the `HTTPClientUpgradeHandler` will create and send the upgrade request. 
+/// This upgrader also assumes that the `HTTPClientUpgradeHandler` will appropriately mutate the
+/// pipeline to remove the HTTP `ChannelHandler`s.
+public final class NIOWebClientSocketUpgrader: NIOHTTPClientProtocolUpgrader {
+    
+    /// RFC 6455 specs this as the required entry in the Upgrade header.
+    public var supportedProtocol: String = "websocket"
+    /// None of the websocket headers are actually defined as 'required'.
+    public var requiredUpgradeHeaders: [String] = []
+    
+    private let requestKey: String
+    private let maxFrameSize: Int
+    private let automaticErrorHandling: Bool
+    private let upgradePipelineHandler: (Channel, HTTPResponseHead) -> EventLoopFuture<Void>
+    
+    public init(requestKey: String,
+                maxFrameSize: Int = 1 << 14,
+                automaticErrorHandling: Bool = true,
+                upgradePipelineHandler: @escaping (Channel, HTTPResponseHead) -> EventLoopFuture<Void>) {
+
+        precondition(requestKey != "", "The request key must contain a valid Sec-WebSocket-Key")
+        precondition(maxFrameSize <= UInt32.max, "invalid overlarge max frame size")
+        self.requestKey = requestKey
+        self.upgradePipelineHandler = upgradePipelineHandler
+        self.maxFrameSize = maxFrameSize
+        self.automaticErrorHandling = automaticErrorHandling
+    }
+
+    public func addCustom(upgradeRequestHeaders: inout HTTPHeaders) {
+        upgradeRequestHeaders.add(name: "Sec-WebSocket-Key", value: self.requestKey)
+        upgradeRequestHeaders.add(name: "Sec-WebSocket-Version", value: "13")
+    }
+
+    public func shouldAllowUpgrade(upgradeResponse: HTTPResponseHead) -> Bool {
```

Comment:
this isn't used or tested but `public`

Reply:
This is part of the conformance to NIOHTTPClientProtocolUpgrader. It is tested as part of the end to end tests. 'testRejectUpgradeIfIncorrectAcceptKey' will provide an incorrect key to this method and check that the upgrade is denied. I added some comments to make it a bit clearer. Again, I can rename on the protocol upgrader?

Reply:
Thanks, this is all good! A comment would be great though. We can't rename because that would break the API.